sdk/core/azure-core/src/main/java/com/azure/android/core/util/logging/AndroidClientLogger.java

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.core.util.logging;
+
+import android.util.Log;
+
+import androidx.annotation.IntDef;
+
+import org.slf4j.LoggerFactory;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+/**
+ * This is a fluent logger helper class that implements the {@link ClientLogger} interface.
+ *
+ * <p>This logger logs formattable messages that use {@code {}} as the placeholder. When a {@link Throwable throwable}
+ * is the last argument of the format varargs and the logger is enabled for
+ * {@link AndroidClientLogger#debug(String) debug}, the stack trace for the throwable is logged.</p>
+ *
+ * <p><strong>Log level hierarchy</strong></p>
+ * <ol>
+ * <li>{@link AndroidClientLogger#error(String) Error}</li>
+ * <li>{@link AndroidClientLogger#warning(String) Warning}</li>
+ * <li>{@link AndroidClientLogger#info(String) Info}</li>
+ * <li>{@link AndroidClientLogger#debug(String) Verbose}</li>
+ * </ol>
+ */
+public class AndroidClientLogger implements ClientLogger {
+    /**
+     * This interface represents the logging levels used in Azure SDKs.
+     */
+    @Retention(RetentionPolicy.SOURCE)
+    @IntDef({DEBUG, INFO, WARNING, ERROR, NOT_SET})
+    public @interface LogLevel {}
+
+    /**
+     * Indicates that log level is at the debug level.
+     */
+    public static final int DEBUG = 1;
+
+    /**
+     * Indicates that log level is at the informational level.
+     */
+    public static final int INFO = 2;
+
+    /**
+     * Indicates that log level is at the warning level.
+     */
+    public static final int WARNING = 3;
+
+    /**
+     * Indicates that log level is at the error level.
+     */
+    public static final int ERROR = 4;
+
+    /**
+     * Indicates that no log level is set.
+     */
+    public static final int NOT_SET = 5;
+
+    private final String tag;
+    private int logLevel;
+
+    /**
+     * Retrieves a logger for the passed class using the {@link LoggerFactory}.
+     *
+     * @param clazz Class creating the logger.
+     */
+    public AndroidClientLogger(Class<?> clazz) {
+        this(clazz.getName());
+    }
+
+    /**
+     * Retrieves a logger for the passed class name using the {@link LoggerFactory}.
+     *
+     * @param tag Class name creating the logger.
+     */
+    public AndroidClientLogger(String tag) {
+        this.tag = tag;
+    }
+
+    /**
+     * Returns this logger's log level.
+     *
+     * @return The log level.
+     */
+    @LogLevel
+    public int getLogLevel() {
```

Comment:
The [documentation for IntDef](https://developer.android.com/reference/androidx/annotation/IntDef) shows this is the correct syntax. Here are some [examples](https://gist.github.com/utkan/bd8cc3d35881e081d9a5c297b9e21044) I found somewhere else.

Reply:
Maybe what you propose also works, I haven't tried it though.

Reply:
Yep, quite right. Looked up the spec and that's the right place - because we have to annotate the method itself, since there is a way to annotate a method parameter but no way to annotate a method return type.